### PR TITLE
Look for code in _GET or _POST explicitly.

### PR DESCRIPTION
**Pull request by SeanZicari**
This change is for those using frameworks or PHP configurations that strip out _REQUEST params.

### CODE REVIEW

**Comment by SeanZicari** [comment]
@glena I'll look into this and see why it failed.

**Comment by glena** [comment]
@SeanZicari dont worry. not mocking the API requests was a bad idea. I have to mock them to get over the rate limit.

**Comment by fgilio** [comment]
For me this is resulting in the following error that was non existent before:

```log
ErrorException in Oauth2Client.php line 240:
Undefined index: code
in Oauth2Client.php line 240
at HandleExceptions->handleError('8', 'Undefined index: code', '/home/vagrant/project/vendor/auth0/auth0-php/src/API/Oauth2Client.php', '240', array()) in Oauth2Client.php line 240
at Oauth2Client->exchangeCode() in Oauth2Client.php line 395
at Oauth2Client->getAccessToken() in Auth0Service.php line 65
at Auth0Service->getUser() in CustomAuth0UserRepository.php line 56
at CustomAuth0UserRepository->getUserByIdentifier('1') in Auth0UserProvider.php line 42
at Auth0UserProvider->retrieveByID('1') in SessionGuard.php line 132
at SessionGuard->user() in GuardHelpers.php line 35
at SessionGuard->authenticate() in AuthManager.php line 294
at AuthManager->__call('authenticate', array()) in Authenticate.php line 57
at Authenticate->authenticate(array()) in Authenticate.php line 41
at Authenticate->handle(object(Request), object(Closure)) in Pipeline.php line 137
at Pipeline->Illuminate\Pipeline\{closure}(object(Request)) in Pipeline.php line 33
at Pipeline->Illuminate\Routing\{closure}(object(Request)) in VerifyCsrfToken.php line 65
at VerifyCsrfToken->handle(object(Request), object(Closure)) in Pipeline.php line 137
at Pipeline->Illuminate\Pipeline\{closure}(object(Request)) in Pipeline.php line 33
at Pipeline->Illuminate\Routing\{closure}(object(Request)) in ShareErrorsFromSession.php line 49
at ShareErrorsFromSession->handle(object(Request), object(Closure)) in Pipeline.php line 137
at Pipeline->Illuminate\Pipeline\{closure}(object(Request)) in Pipeline.php line 33
at Pipeline->Illuminate\Routing\{closure}(object(Request)) in StartSession.php line 64
at StartSession->handle(object(Request), object(Closure)) in Pipeline.php line 137
at Pipeline->Illuminate\Pipeline\{closure}(object(Request)) in Pipeline.php line 33
at Pipeline->Illuminate\Routing\{closure}(object(Request)) in AddQueuedCookiesToResponse.php line 37
at AddQueuedCookiesToResponse->handle(object(Request), object(Closure)) in Pipeline.php line 137
at Pipeline->Illuminate\Pipeline\{closure}(object(Request)) in Pipeline.php line 33
at Pipeline->Illuminate\Routing\{closure}(object(Request)) in EncryptCookies.php line 59
at EncryptCookies->handle(object(Request), object(Closure)) in Pipeline.php line 137
at Pipeline->Illuminate\Pipeline\{closure}(object(Request)) in Pipeline.php line 33
at Pipeline->Illuminate\Routing\{closure}(object(Request)) in Pipeline.php line 104
at Pipeline->then(object(Closure)) in Router.php line 655
at Router->runRouteWithinStack(object(Route), object(Request)) in Router.php line 629
at Router->dispatchToRoute(object(Request)) in Router.php line 607
at Router->dispatch(object(Request)) in Kernel.php line 268
at Kernel->Illuminate\Foundation\Http\{closure}(object(Request)) in Pipeline.php line 53
at Pipeline->Illuminate\Routing\{closure}(object(Request)) in Debugbar.php line 51
at Debugbar->handle(object(Request), object(Closure)) in Pipeline.php line 137
at Pipeline->Illuminate\Pipeline\{closure}(object(Request)) in Pipeline.php line 33
at Pipeline->Illuminate\Routing\{closure}(object(Request)) in CheckForMaintenanceMode.php line 46
at CheckForMaintenanceMode->handle(object(Request), object(Closure)) in Pipeline.php line 137
at Pipeline->Illuminate\Pipeline\{closure}(object(Request)) in Pipeline.php line 33
at Pipeline->Illuminate\Routing\{closure}(object(Request)) in Pipeline.php line 104
at Pipeline->then(object(Closure)) in Kernel.php line 150
at Kernel->sendRequestThroughRouter(object(Request)) in Kernel.php line 117
at Kernel->handle(object(Request)) in index.php line 53
```

Had to shortcircuit the function with a ```return;```

Am I the only one?

**Comment by glena** [comment]
@fgilio upgrade to 4.0.10 and try again, it was wrong the check on GET and POST

**Comment by fgilio** [comment]
Ok, I'll let you know. Thanks!

File: src/API/Oauth2Client.php [hunk 1]
```diff
@@ -237,10 +237,11 @@ private function dontPersist($name) {
      * @return Boolean Whether it exchanged the code or not correctly
      */
     public function exchangeCode() {
-        if (!isset($_REQUEST['code'])) {
+        $code = $_GET['code'] ?: $_POST['code'];
```
Comment (glena) [inline src/API/Oauth2Client.php]: it should be `$code = isset($_GET['code']) ? $_GET['code'] : $_POST['code'];` 
or `$code = $_GET['code'] ?? $_POST['code'];`

Comment (fgilio) [inline src/API/Oauth2Client.php]: What about ```$code = isset($_GET['code']) ? $_GET['code'] : ( isset($_POST['code']) ? $_POST['code'] : null );```

Comment (fgilio) [inline src/API/Oauth2Client.php]: *Tried those two, but got the same error ```Undefined index: code```

Comment (glena) [inline src/API/Oauth2Client.php]: but there is a GET or POST param `code`?

Comment (fgilio) [inline src/API/Oauth2Client.php]: In my case, $_GET['code'] nor $_POST['code'] seem to be set

Comment (glena) [inline src/API/Oauth2Client.php]: try 4.0.11

Comment (fgilio) [inline src/API/Oauth2Client.php]: Working fine, and much much more readable than what I said 😅

Comment (glena) [inline src/API/Oauth2Client.php]: hahaha sorry for this issue. when I review the PR misread `?:` by `??: :(

thx for reporting

Comment (fgilio) [inline src/API/Oauth2Client.php]: Haha no problem, thank you for fixing it so fast! 👍🏼

**Comment by SeanZicari** [comment]
I'm not sure what version of PHP you're using, but ?: works fine for me. As far as I know, that is the equivalent of JavaScript's 

var1 || var2

syntax

**Comment by SeanZicari** [comment]
Ah, I see. ?? Is the null coalesce operator. However, ?: is the short ternary operator and should also have worked. I guess it's not as good as the null coalesce operator, though.

**Pull request merged**